recommendations and guidelines section (#128)
huggingface/distil-whisper: training/README.md

print("Environment set up successful?", generated_ids.shape[-1] == 20)
```

> ## 0. Recommendations and guidelines 

Reviewer: Should we put this section at the end of the README? i.e. before https://github.com/huggingface/distil-whisper/tree/main/training#overview-of-training-methods

If we do this, we can assume the user has read the entire README once end-to-end, and so the section content will make more sense

Author: Yes, I agree, this way mentions to pseudo labelling etc will be clearer!

### 0.1 Datasets

> As explained, ideally, you should aim for ~1000 hours of audio data for training a distilled model via KD. Moreover, you should use quality out-of-distribution test sets to assess generalization capacities. With at least 1500 hours of audio data for German, Dutch, French and Spanish, 600 hours for Italian, and 300 hours for Portuguese and Polish (which can be supplemented with your own datasets), a good setup to start with is:

Reviewer: ```suggestion
As explained, ideally, you should aim for ~1000 hours of audio data for training a distilled model via KD. Moreover, you should evaluate your model on out-of-distribution test sets to assess generalization capacities. With at least 1500 hours of audio data for German, Dutch, French and Spanish, 600 hours for Italian, and 300 hours for Portuguese and Polish (which can be supplemented with your own datasets), a good setup to start with is:
```

> - [Common Voice 17](https://huggingface.co/datasets/mozilla-foundation/common_voice_17_0) and [Multilingual Librispeech](https://huggingface.co/datasets/facebook/multilingual_librispeech): **train slipts**, **val splits** for in training evaluation, **test splits** for in-distribution testing 
> - [VoxPopuli](https://huggingface.co/datasets/facebook/voxpopuli) and [Fleurs](https://huggingface.co/datasets/google/fleurs): **test splits** for out-of-distribution testing

Reviewer: ```suggestion
- **Training datasets:** [Common Voice 17](https://huggingface.co/datasets/mozilla-foundation/common_voice_17_0) and [Multilingual Librispeech](https://huggingface.co/datasets/facebook/multilingual_librispeech). Use the `train` split for training, and the `validation` and `test` splits for in-distribution testing.
- **Test datasets:** [VoxPopuli](https://huggingface.co/datasets/facebook/voxpopuli) and [Fleurs](https://huggingface.co/datasets/google/fleurs). Use the `validation` and `test` splits for out-of-distribution testing.
```

### 0.2 Student model's decoder 

> **Number of decoder layers:**  We recommend using a 2-layers decoder (see language transfer below).  However, you can adjust the number of decoder layers when initializing the student model to balance between inference speed and accuracy. Experimentation has revealed that the Pareto optimal points are with 2, 3, and 4-layers decoders. For indicative results, after 10,000 training steps and inference on an 80GB Nvidia H100 with a batch size of 16, compared to a 2-layers decoder:

Reviewer: If we use section headings then we can reference the sub-sections in the rest of the README

```suggestion
#### 0.2.1 Number of Decoder Layers

We recommend using a 2-layers decoder (see language transfer below).  However, you can adjust the number of decoder layers when initializing the student model to balance between inference speed and accuracy. Experimentation has revealed that the Pareto optimal points are with 2, 3, and 4-layers decoders. For indicative results, after 10,000 training steps and inference on an 80GB Nvidia H100 with a batch size of 16, compared to a 2-layers decoder:
```

<center>

> |    |  rel. token gen. speed |  ΔWER(%) |

Reviewer: Rather than doing this relative to 2-layers, I would include the 2-layer results in the table, and do everything relative to the teacher model. This way users can get a feel for how much faster each variant will be relative to the baseline

|----------|:-------------:|------:|
| 3 layers |  $0.91$ | $-1.2$ |
| 4 layers |    $0.85$   |   $-1.7$ |

</center>

> **Language transfer (for a 2-layers decoder):** If you opt for a 2-layers decoder, consider leveraging language transfer by initializing the student model from the [distil-large-v3 English distilled model](https://huggingface.co/distil-whisper/distil-large-v3). For French, this method has shown performance improvements of ΔWER=-1.9% (compared to a 2-layers decoder initialized from [Whiper *large-v3*](https://huggingface.co/openai/whisper-large-v3)) after 10,000 training steps.

Reviewer: ```suggestion
#### 0.2.2 Language Transfer

If you opt for a 2-layers decoder, consider leveraging language transfer by initializing the student model from the [distil-large-v3 English distilled model](https://huggingface.co/distil-whisper/distil-large-v3). For French, this method has shown performance improvements of ΔWER=-1.9% (compared to a 2-layers decoder initialized from [Whiper *large-v3*](https://huggingface.co/openai/whisper-large-v3)) after 10,000 training steps.
```

### 0.3 Language mixing

> If you're working with low-resource languages (<500 hours of audio data), consider mixing your training data with a closely related language (for example, mix French and Spanish). To do this:

Reviewer: ```suggestion
If you're working with low-resource languages (<500 hours of audio data), consider mixing your training data with a closely related language (for example, mix French and Spanish) to leverage knowledge transfer between languages. To do this:
```

1. Run [pseudo labeling](#1-pseudo-labelling) for each training dataset, setting the `--language` flag to the language of the respective dataset (e.g., French and Spanish).

> 2. Conduct training on these pseudo-labeled datasets, using the `--language` flag set to your targeted language (e.g., French). Note that this flag is only used for evaluation purposes, so you set it to the targeted language. The language token used for forwarding the teacher and student model decoders is the one used and saved in pseudo labels during pseudo-labeling, ensuring it's the correct one for the considered sample.

Reviewer: You can give an example of how the arguments change for language mixing:
```diff
-  --train_dataset_name "../common_voice_17_0_fr_pseudo_labelled \
-  --train_split_name "train" \
+  --train_dataset_name "../common_voice_17_0_fr_pseudo_labelled+../common_voice_17_0_es_pseudo_labelled" \
+  --train_split_name "train+train" \
```

1. `freeze_encoder`: whether to freeze the entire encoder of the student model during training. Beneficial when the student encoder is copied exactly from the teacher encoder. In this case, the encoder hidden-states from the teacher model are re-used for the student model. Stopping the gradient computation through the encoder and sharing the encoder hidden-states provides a significant memory saving, and can enable up to 2x batch sizes. 
2. `freeze_embed_positions`: whether to freeze the student model's decoder positional embeddings. Using the same embed positions as the teacher model, which is designed to handle context lengths up to 448 tokens, helps the student model retain its input id representation up to the full max input length. 
3. `dtype`: data type (dtype) in which the model computation should be performed. Note that this only controls the dtype of the computations (forward and backward pass), and not the dtype of the parameters or optimiser states.

> 4. `freeze_decoder`: wether to freeze the student model's decoder. Note that the input tokens embeddings and language modelling head will remain trainable.  

Reviewer: ```suggestion
4. `freeze_decoder`: whether to freeze the student model's decoder. Note that the input tokens embeddings and language modelling head will remain trainable.  
```
